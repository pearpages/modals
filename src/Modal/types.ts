export type ModalId = string;
export type ModalSize = 'sm' | 'md' | 'lg' | 'xl' | { maxWidth: number };
export type CloseReason = 'esc' | 'backdrop' | 'button' | 'programmatic' | 'unmount' | 'route-change';

export interface ModalOptions {
  id?: ModalId;
  title?: string;
  role?: 'dialog' | 'alertdialog';
  size?: ModalSize;
  ariaLabel?: string;
  labelledBy?: string;
  describedBy?: string;
  closeOnEsc?: boolean;
  closeOnBackdrop?: boolean;
  initialFocus?: HTMLElement | null;
  returnFocus?: HTMLElement | null;
  trapFocus?: boolean;
  inertBackground?: boolean;
  preventScroll?: boolean;
  portalTarget?: HTMLElement;
  onClose?: (reason: CloseReason) => void;
  onOpen?: () => void;
  animations?: {
    enter?: string;
    exit?: string;
    backdropEnter?: string;
    backdropExit?: string;
    durationMs?: number;
  };
}

export interface OpenPayload extends ModalOptions {
  content: React.ReactNode | ((ctx: { close: () => void; id: ModalId }) => React.ReactNode);
}

export interface ModalEntry extends OpenPayload {
  id: ModalId;
  state: 'open' | 'closing';
  openedAt: number;
}

export interface ModalApi {
  open: (payload: OpenPayload) => ModalId;
  close: (id?: ModalId, reason?: CloseReason) => void;
  replace: (id: ModalId, next: OpenPayload) => void;
  update: (id: ModalId, patch: Partial<OpenPayload>) => void;
  isOpen: (id: ModalId) => boolean;
  subscribe: (fn: (stack: ModalEntry[]) => void) => () => void;
}
