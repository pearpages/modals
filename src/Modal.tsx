import React, { useEffect, useCallback } from 'react';
import { ModalProps } from './types';
import { useModalContext } from './ModalProvider';
import { ModalIdProvider } from './ModalIdContext';
import { ModalTrigger } from './ModalTrigger';
import { ModalContent } from './ModalContent';
import { ModalHeader } from './ModalHeader';
import { ModalTitle } from './ModalTitle';
import { ModalDescription } from './ModalDescription';
import { ModalClose } from './ModalClose';
import { ModalFooter } from './ModalFooter';

// Type for compound component
interface ModalComponent extends React.FC<ModalProps> {
  Trigger: typeof ModalTrigger;
  Content: typeof ModalContent;
  Header: typeof ModalHeader;
  Title: typeof ModalTitle;
  Description: typeof ModalDescription;
  Close: typeof ModalClose;
  Footer: typeof ModalFooter;
}

/**
 * Modal component that supports controlled state management.
 * 
 * Features:
 * - Controlled via `open` prop and `onOpenChange` callback
 * - Automatically registers/unregisters with ModalProvider
 * - Unmounts children when closed (modal content is not rendered when closed)
 * - Integrates with modal stack for proper z-index and focus management
 */
const Modal: ModalComponent = ({ 
  id, 
  open, 
  onOpenChange, 
  children, 
  className 
}) => {
  const modalContext = useModalContext();

  const { 
    register, 
    unregister, 
    openModal, 
    closeModal, 
    updateOnOpenChange,
    getModalEntry 
  } = modalContext;

  // Register modal on mount, unregister on unmount
  useEffect(() => {
    register(id);
    return () => {
      unregister(id);
    };
  }, [id, register, unregister]);

  // Update onOpenChange callback in registry
  useEffect(() => {
    updateOnOpenChange(id, onOpenChange);
  }, [id, onOpenChange, updateOnOpenChange]);

  // Get current modal state from provider
  const modalEntry = getModalEntry(id);
  const isProviderOpen = modalEntry?.open ?? false;

  // Sync controlled prop with provider state
  useEffect(() => {
    if (open !== undefined && open !== isProviderOpen) {
      if (open) {
        openModal(id);
      } else {
        closeModal(id);
      }
    }
  }, [open, isProviderOpen, id, openModal, closeModal]);

  // Notify parent of provider state changes (for uncontrolled usage or external changes)
  const prevProviderOpenRef = React.useRef(isProviderOpen);
  useEffect(() => {
    const prevProviderOpen = prevProviderOpenRef.current;
    prevProviderOpenRef.current = isProviderOpen;
    
    // Notify if provider state changed
    if (onOpenChange && isProviderOpen !== prevProviderOpen) {
      // In uncontrolled mode (open === undefined), always notify
      // In controlled mode, notify if the provider state differs from the controlled prop
      // This handles external changes like backdrop clicks or escape key
      if (open === undefined || isProviderOpen !== open) {
        onOpenChange(isProviderOpen);
      }
    }
  }, [isProviderOpen, onOpenChange, open]);

  // Handle controlled close via onOpenChange
  const handleClose = useCallback(() => {
    if (onOpenChange) {
      onOpenChange(false);
    } else {
      closeModal(id);
    }
  }, [onOpenChange, closeModal, id]);

  // Only render children when modal is open (unmount when closed)
  if (!isProviderOpen) {
    return null;
  }

  return (
    <ModalIdProvider modalId={id}>
      <div className={className} data-modal-id={id}>
        {children}
      </div>
    </ModalIdProvider>
  );
};

Modal.displayName = 'Modal';

// Compound component pattern
Modal.Trigger = ModalTrigger;
Modal.Content = ModalContent;
Modal.Header = ModalHeader;
Modal.Title = ModalTitle;
Modal.Description = ModalDescription;
Modal.Close = ModalClose;
Modal.Footer = ModalFooter;

export { Modal };
