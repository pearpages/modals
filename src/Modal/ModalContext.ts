import { createContext } from 'react';
import type { ModalApi } from './types';

export const ModalContext = createContext<ModalApi | null>(null);
