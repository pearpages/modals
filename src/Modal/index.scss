/*
 * Modal Component Styles
 * 
 * Responsive Behavior:
 * - Mobile (â‰¤768px): Full-screen modals with slide-up animation
 * - Desktop (>768px): Centered modals with different size variants
 * 
 * Features:
 * - Safe area handling for devices with notches
 * - Smooth animations and transitions
 * - Accessibility-focused design
 * - Proper z-index stacking for multiple modals
 */

// Modal theme variables
:root {
  --modal-bg: #ffffff;
  --modal-text: #333333;
  --modal-border: #e5e7eb;
  --modal-button-bg: #ffffff;
  --modal-button-hover: #f3f4f6;
  --modal-button-text: #374151;
  --modal-button-border: #d1d5db;
  --modal-actions-bg: #f9fafb;
  --modal-close-text: #6b7280;
  --modal-close-hover: #374151;
}

@media (prefers-color-scheme: dark) {
  :root {
    --modal-bg: #1e1e1e;
    --modal-text: #e0e0e0;
    --modal-border: #333333;
    --modal-button-bg: #2d2d2d;
    --modal-button-hover: #3d3d3d;
    --modal-button-text: #e0e0e0;
    --modal-button-border: #404040;
    --modal-actions-bg: #252525;
    --modal-close-text: #9ca3af;
    --modal-close-hover: #d1d5db;
  }
}

// Modal backdrop - uses high specificity instead of !important
[data-modal-root].modal-backdrop {
  // Positioning
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  // Note: Using top/left/right/bottom is more reliable than width/height on mobile Safari
  // width: 100vw and height: 100vh can cause issues with Safari's dynamic viewport
  
  // Background and layout
  background-color: rgba(0, 0, 0, 0.5);
  
  // Flexbox layout
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 0;
  margin: 0;
  
  // Ensure it's above everything
  z-index: 999999;
  
  // Animation states - backdrop fade only
  opacity: 0;
  transition: opacity 250ms ease-out;
  
  &[data-state="open"] {
    opacity: 1;
  }
  
  &[data-state="closing"] {
    opacity: 0;
  }
  
  // Stacked modals should not have dark backdrop
  &.modal-backdrop--stacked {
    background-color: transparent;
  }
}

// Modal dialog container - uses high specificity instead of !important
[data-modal-root] .modal-dialog {
  // Reset all possible conflicting styles
  position: relative;
  margin: 0;
  padding: 0;
  
  // Layout
  background: var(--modal-bg, #ffffff);
  border-radius: 8px;
  box-shadow: 0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04);
  
  // Size constraints for desktop
  width: auto;
  min-width: 320px;
  max-width: 600px;
  max-height: 90vh;
  
  // Flexbox layout
  display: flex;
  flex-direction: column;
  
  // Animation states - transform only, no opacity conflicts
  transform: scale(0.95) translateY(-10px);
  transition: transform 250ms cubic-bezier(0.16, 1, 0.3, 1);
  
  &[data-state="open"] {
    transform: scale(1) translateY(0);
  }
  
  &[data-state="closing"] {
    transform: scale(0.95) translateY(-10px);
  }
}

// Modal header
.modal-header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 1.5rem;
  border-bottom: 1px solid var(--modal-border, #e5e7eb);
  flex-shrink: 0; // Prevent header from shrinking
  
  .modal-title {
    margin: 0;
    font-size: 1.25rem;
    font-weight: 600;
    color: var(--modal-text, #111827);
  }
  
  .modal-close {
    background: none;
    border: none;
    font-size: 1.5rem;
    cursor: pointer;
    padding: 0.25rem;
    border-radius: 4px;
    color: var(--modal-close-text, #6b7280);
    line-height: 1;
    min-width: 2rem;
    min-height: 2rem;
    display: flex;
    align-items: center;
    justify-content: center;
    
    &:hover {
      background-color: var(--modal-button-hover, #f3f4f6);
      color: var(--modal-close-hover, #374151);
    }
    
    &:focus {
      outline: 2px solid #3b82f6;
      outline-offset: 2px;
    }
  }
}

// Modal content
.modal-content {
  padding: 1.5rem;
  flex: 1;
  overflow-y: auto;
}

// Modal without header
.modal-dialog:not(:has(.modal-header)) .modal-content {
  padding-top: 1.5rem;
}

// Confirm modal actions
.modal-confirm-actions {
  display: flex;
  gap: 0.75rem;
  justify-content: flex-end;
  margin-top: 1.5rem;
  
  button {
    padding: 0.5rem 1rem;
    border-radius: 6px;
    border: 1px solid var(--modal-button-border, #d1d5db);
    background: var(--modal-button-bg, white);
    color: var(--modal-button-text, #374151);
    cursor: pointer;
    font-weight: 500;
    transition: all 150ms ease;
    
    &:hover {
      background-color: var(--modal-button-hover, #f9fafb);
    }
    
    &:focus {
      outline: 2px solid #3b82f6;
      outline-offset: 2px;
    }
    
    &[data-variant="danger"] {
      background-color: #dc2626;
      border-color: #dc2626;
      color: white;
      
      &:hover {
        background-color: #b91c1c;
      }
    }
    
    &:last-child {
      background-color: #3b82f6;
      border-color: #3b82f6;
      color: white;
      
      &:hover {
        background-color: #2563eb;
      }
    }
  }
}

// Focus trap container
[data-modal-root] div[style*="outline: none"] {
  outline: none;
}

// Modal focus trap container - ensures full area is clickable
.modal-focus-trap {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  // Note: Using 100% instead of viewport units since parent (backdrop) handles viewport sizing
  outline: none;
  
  // Safari-specific fixes for margin issues
  @supports (-webkit-touch-callout: none) {
    // iOS Safari specific
    position: relative;
    -webkit-overflow-scrolling: touch;
  }
}

// Mobile responsive design - full screen
@media (max-width: 768px) {
  [data-modal-root].modal-backdrop {
    padding: 0;
    align-items: stretch;
    justify-content: stretch;
  }
  
  [data-modal-root] .modal-dialog {
    width: 100vw;
    height: 100vh;
    height: 100dvh; // Use dynamic viewport height if supported
    max-width: none;
    max-height: none;
    border-radius: 0;
    
    // Safe area handling for devices with notches
    padding-top: env(safe-area-inset-top, 0);
    padding-bottom: env(safe-area-inset-bottom, 0);
    
    // Full screen slide-up animation for mobile - transform only
    transform: translateY(100%);
    transition: transform 300ms cubic-bezier(0.16, 1, 0.3, 1);
    
    &[data-state="open"] {
      transform: translateY(0);
    }
    
    &[data-state="closing"] {
      transform: translateY(100%);
    }
    
    // Override all size variants on mobile
    &[data-size="sm"],
    &[data-size="md"], 
    &[data-size="lg"],
    &[data-size="xl"],
    &.modal-sm,
    &.modal-md,
    &.modal-lg,
    &.modal-xl {
      width: 100vw;
      max-width: none;
    }
  }
  
  [data-modal-root] .modal-header {
    padding: 1rem;
    padding-top: calc(1rem + env(safe-area-inset-top, 0));
    border-bottom: 1px solid var(--modal-border, #e5e7eb);
    flex-shrink: 0;
  }
  
  [data-modal-root] .modal-content {
    padding: 1rem;
    padding-bottom: calc(1rem + env(safe-area-inset-bottom, 0));
    flex: 1;
    overflow-y: auto;
    -webkit-overflow-scrolling: touch; // Smooth scrolling on iOS
  }
}

// Desktop styles - ensure proper centering
@media (min-width: 769px) {
  [data-modal-root].modal-backdrop {
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 2rem;
  }
  
  [data-modal-root] .modal-dialog {
    margin: 0;
    width: auto;
    height: auto;
    max-height: 90vh;
    border-radius: 8px;
    
    // Size variants with proper max-widths for desktop
    &[data-size="sm"], &.modal-sm {
      width: 100%;
      max-width: 400px;
    }
    
    &[data-size="md"], &.modal-md {
      width: 100%;
      max-width: 600px;
    }
    
    &[data-size="lg"], &.modal-lg {
      width: 100%;
      max-width: 800px;
    }
    
    &[data-size="xl"], &.modal-xl {
      width: 100%;
      max-width: 1200px;
    }
  }
}

// Modal action utilities for common button layouts
.modal-actions {
  display: flex;
  gap: 0.75rem;
  padding: 1.5rem;
  border-top: 1px solid var(--modal-border, #e5e7eb);
  background-color: var(--modal-actions-bg, #f9fafb);
  flex-shrink: 0;
  
  // Right-aligned actions (common pattern)
  &.modal-actions-right {
    justify-content: flex-end;
  }
  
  // Center-aligned actions
  &.modal-actions-center {
    justify-content: center;
  }
  
  // Space between actions (first left, rest right)
  &.modal-actions-between {
    justify-content: space-between;
  }
  
  // Full-width single action
  &.modal-actions-full {
    button {
      flex: 1;
    }
  }
}

// Button styles for modal actions
.modal-actions {
  button {
    padding: 0.5rem 1rem;
    border-radius: 6px;
    border: 1px solid var(--modal-button-border, #d1d5db);
    background-color: var(--modal-button-bg, white);
    color: var(--modal-button-text, #374151);
    font-weight: 500;
    cursor: pointer;
    transition: all 150ms ease-in-out;
    
    &:hover {
      background-color: var(--modal-button-hover, #f3f4f6);
      border-color: var(--modal-button-border, #9ca3af);
    }
    
    &:focus {
      outline: 2px solid #3b82f6;
      outline-offset: 2px;
    }
    
    // Primary button variant
    &.btn-primary {
      background-color: #3b82f6;
      border-color: #3b82f6;
      color: white;
      
      &:hover {
        background-color: #2563eb;
        border-color: #2563eb;
      }
    }
    
    // Danger button variant
    &.btn-danger {
      background-color: #ef4444;
      border-color: #ef4444;
      color: white;
      
      &:hover {
        background-color: #dc2626;
        border-color: #dc2626;
      }
    }
  }
}

// Responsive modal actions
@media (max-width: 768px) {
  .modal-actions {
    padding: 1rem;
    padding-bottom: calc(1rem + env(safe-area-inset-bottom, 0));
    
    // Stack buttons on mobile for better touch targets
    &.modal-actions-mobile-stack {
      flex-direction: column;
      
      button {
        width: 100%;
        min-height: 44px; // Minimum touch target size
      }
    }
  }
}
