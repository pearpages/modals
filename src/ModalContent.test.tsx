import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import React from 'react';
import { Modal } from './Modal';
import { ModalProvider } from './ModalProvider';
import { ModalRoot } from './ModalRoot';

// Test wrapper that provides full modal system context
const TestModalSystem: React.FC<{
  open?: boolean;
  onOpenChange?: (open: boolean) => void;
  id?: string;
  children?: React.ReactNode;
}> = ({ open = false, onOpenChange, id = 'test-modal', children }) => {
  return (
    <ModalProvider>
      <ModalRoot />
      <Modal id={id} open={open} onOpenChange={onOpenChange}>
        {children}
      </Modal>
    </ModalProvider>
  );
};

// Helper function to render with proper act handling and timer control
const renderModal = async (ui: React.ReactElement) => {
  let result: any;
  await act(async () => {
    result = render(ui);
    // Fast-forward any timers to complete state transitions immediately
    if (vi.isMockFunction(setTimeout)) {
      vi.runAllTimers();
    }
  });
  return result;
};

// Helper for snapshot tests
const renderModalSnapshot = async (ui: React.ReactElement) => {
  let result: any;
  await act(async () => {
    result = render(ui);
    // Fast-forward timers for consistent snapshots
    if (vi.isMockFunction(setTimeout)) {
      vi.runAllTimers();
    }
  });
  return result;
};

describe('Modal.Content - Layout and Sizes', () => {
  beforeEach(() => {
    // Use fake timers for precise control over async state updates
    vi.useFakeTimers();
    
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  afterEach(() => {
    // Restore real timers after each test
    vi.useRealTimers();
  });

  it('should render with default size (md)', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).toContain('modal');
    expect(modalElement?.className).toContain('modal--md');
  });

  it('should apply auto size class', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content size="auto">
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).toContain('modal--auto');
  });

  it('should apply full size class', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content size="full">
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).toContain('modal--full');
  });

  it('should apply custom className', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content className="custom-modal">
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).toContain('custom-modal');
  });

  it('should not render when modal is closed', async () => {
    await renderModal(
      <TestModalSystem open={false}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    expect(screen.queryByTestId('content')).toBe(null);
  });

  it('should apply full size class', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content size="full">
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement).toHaveClass('modal--full');
  });

  it('should apply custom className', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content className="custom-modal">
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement).toHaveClass('custom-modal');
  });

  it('should not render when modal is closed', () => {
    render(
      <TestModalSystem open={false}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    expect(screen.queryByTestId('content')).toBe(null);
  });
});

describe('Modal.Content - Animation and Data States', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should have animated class when animated=true (default)', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).toContain('modal--animated');
  });

  it('should not have animated class when animated=false', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content animated={false}>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.className).not.toContain('modal--animated');
  });

  it('should set data-state attribute for animations', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement?.getAttribute('data-state')).toBeTruthy();
    
    // Should be either 'opening' or 'open'
    const dataState = modalElement?.getAttribute('data-state');
    expect(['opening', 'open']).toContain(dataState);
  });

  it('should set data-state on backdrop', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    // The backdrop is now handled by ModalRoot, check for the portal backdrop
    const backdrop = document.querySelector('[data-modal-backdrop="true"]');
    expect(backdrop).toBeTruthy();
    // ModalRoot backdrop doesn't use data-state, it handles styling differently
  });
});

describe('Modal.Content - Accessibility', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should have role="dialog"', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement).toHaveAttribute('role', 'dialog');
  });

  it('should have aria-modal="true"', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    const modalElement = content.closest('[role="dialog"]');
    expect(modalElement).toHaveAttribute('aria-modal', 'true');
  });
});

describe('Modal.Content - Dismiss Behavior', () => {
  beforeEach(() => {
    // Don't use fake timers for dismiss behavior tests since they use waitFor
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  it('should close on backdrop click by default', async () => {
    const onOpenChange = vi.fn();
    await renderModal(
      <TestModalSystem open={true} onOpenChange={onOpenChange}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    // Find the backdrop element created by ModalRoot (has data-modal-backdrop="true")
    const backdrop = document.querySelector('[data-modal-backdrop="true"]');
    
    if (backdrop) {
      fireEvent.click(backdrop);
      await waitFor(() => {
        expect(onOpenChange).toHaveBeenCalledWith(false);
      });
    } else {
      throw new Error('Backdrop element not found');
    }
  });

  it('should not close on backdrop click when closeOnBackdrop=false', async () => {
    const onOpenChange = vi.fn();
    await renderModal(
      <TestModalSystem open={true} onOpenChange={onOpenChange}>
        <Modal.Content closeOnBackdrop={false}>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    // Find the backdrop element created by ModalRoot (has data-modal-backdrop="true")
    const backdrop = document.querySelector('[data-modal-backdrop="true"]');
    
    if (backdrop) {
      fireEvent.click(backdrop);
      await waitFor(() => {
        expect(onOpenChange).not.toHaveBeenCalled();
      }, { timeout: 100 });
    }
  });

  it('should not close on content click', async () => {
    const onOpenChange = vi.fn();
    await renderModal(
      <TestModalSystem open={true} onOpenChange={onOpenChange}>
        <Modal.Content>
          <div data-testid="content">Content</div>
        </Modal.Content>
      </TestModalSystem>
    );

    const content = screen.getByTestId('content');
    fireEvent.click(content);
    
    await waitFor(() => {
      expect(onOpenChange).not.toHaveBeenCalled();
    }, { timeout: 100 });
  });
});

describe('Modal.Content - Multiple Sizes Snapshot', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should match snapshot for auto size', async () => {
    const { container } = await renderModalSnapshot(
      <TestModalSystem open={true}>
        <Modal.Content size="auto">
          <Modal.Header>
            <Modal.Title>Auto Size Modal</Modal.Title>
            <Modal.Close />
          </Modal.Header>
          <div>Auto-sized content</div>
        </Modal.Content>
      </TestModalSystem>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for md size', async () => {
    const { container } = await renderModalSnapshot(
      <TestModalSystem open={true}>
        <Modal.Content size="md">
          <Modal.Header>
            <Modal.Title>Medium Modal</Modal.Title>
            <Modal.Close />
          </Modal.Header>
          <div>Medium-sized content</div>
        </Modal.Content>
      </TestModalSystem>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for full size', async () => {
    const { container } = await renderModalSnapshot(
      <TestModalSystem open={true}>
        <Modal.Content size="full">
          <Modal.Header>
            <Modal.Title>Full Size Modal</Modal.Title>
            <Modal.Close />
          </Modal.Header>
          <div>Full-screen content</div>
        </Modal.Content>
      </TestModalSystem>
    );
    
    expect(container).toMatchSnapshot();
  });
});

describe('Modal Subcomponents', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    
    // Mock portal container
    const portalDiv = document.createElement('div');
    portalDiv.setAttribute('id', 'modal-portal');
    document.body.appendChild(portalDiv);
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should render Modal.Header with proper class', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <Modal.Header data-testid="header">
            <Modal.Title>Title</Modal.Title>
          </Modal.Header>
        </Modal.Content>
      </TestModalSystem>
    );

    const header = screen.getByTestId('header');
    expect(header).toHaveClass('modal-header');
  });

  it('should render Modal.Title with auto-generated ID', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <Modal.Title data-testid="title">Test Title</Modal.Title>
        </Modal.Content>
      </TestModalSystem>
    );

    const title = screen.getByTestId('title');
    expect(title).toHaveAttribute('id');
    expect(title.id).toMatch(/modal-title-/);
  });

  it('should render Modal.Description with auto-generated ID', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <Modal.Description data-testid="description">
            Test Description
          </Modal.Description>
        </Modal.Content>
      </TestModalSystem>
    );

    const description = screen.getByTestId('description');
    expect(description).toHaveAttribute('id');
    expect(description.id).toMatch(/modal-description-/);
  });

  it('should render Modal.Close with default close icon', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <Modal.Close data-testid="close" />
        </Modal.Content>
      </TestModalSystem>
    );

    const close = screen.getByTestId('close');
    expect(close).toHaveTextContent('×');
    expect(close).toHaveAttribute('aria-label', 'Close modal');
  });

  it('should render Modal.Footer with proper class', async () => {
    await renderModal(
      <TestModalSystem open={true}>
        <Modal.Content>
          <Modal.Footer data-testid="footer">
            <button>Cancel</button>
            <button>Confirm</button>
          </Modal.Footer>
        </Modal.Content>
      </TestModalSystem>
    );

    const footer = screen.getByTestId('footer');
    expect(footer).toHaveClass('modal-footer');
  });
});