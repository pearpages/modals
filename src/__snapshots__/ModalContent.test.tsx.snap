// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal.Content - Multiple Sizes Snapshot > should match snapshot for auto size 1`] = `
<div>
  <div
    data-modal-id="test-modal"
  />
</div>
`;

exports[`Modal.Content - Multiple Sizes Snapshot > should match snapshot for full size 1`] = `
<div>
  <div
    data-modal-id="test-modal"
  />
</div>
`;

exports[`Modal.Content - Multiple Sizes Snapshot > should match snapshot for md size 1`] = `
<div>
  <div
    data-modal-id="test-modal"
  />
</div>
`;
